gr-digital/lib/costas_loop_cc_impl.cc

```diff
@@ -171,22 +89,46 @@ int costas_loop_cc_impl::work(int noutput_items,
             }
         }
 
-        nco_out = gr_expj(-d_phase);
-        optr[i] = iptr[i] * nco_out;
-
-        d_error = (*this.*d_phase_detector)(optr[i]);
+        const gr_complex nco_out = gr_expj(-d_phase);
+
+        gr::fast_cc_multiply(optr[i], iptr[i], nco_out);
+
+        // EXPENSIVE LINE with function pointer, switch was about 20% faster in testing.
+        // Left in for logic justification/reference. d_error = phase_detector_2(optr[i]);
+        switch (d_order) {
```

Comment:
A moving target here, of sorts! Before you had different `work` implementation, one of which was called based on the instantiation settings. Now it's a `switch` and `if` to determine which is called, inside `work` ... which is really inefficient!

Suggestion: Since the settings can't be changed after instantiation, create a class variable that's a pointer to the method to call based on the instantiation arguments. Then just call it here. Much tighter code & will be faster too!